### PR TITLE
Exception handling rework

### PR DESCRIPTION
**Pull request by elliottpope**
Still to be done
- [ ] Allow users to inject their own success and error handlers to specific `@JMSListeners` using `errorHandlers` and `successHandlers` annotation parameters (error handlers done, success handlers in a different issue?)
- [ ] Allow users to configure global success and error handlers to be injected to all `@JMSListeners` 
- [ ] Implement the `ComplexObject` test cases and examples in Groovy and Java
- [x] Implement test cases and examples for adding custom error and success handling logic

### CODE REVIEW

File: jms-core/src/main/java/io/micronaut/jms/listener/ConcurrentMessageHandler.java [hunk 1]
```diff
@@ -29,7 +29,9 @@
  * @param <T> the type of the object that the handler is expecting.
  * @author Elliott Pope
  * @since 1.0.0
+ * @deprecated since 1.0.0.M2. All associated logic is handled by the {@link JMSListener}.
```
Comment (burtbeckwith) [inline jms-core/src/main/java/io/micronaut/jms/listener/ConcurrentMessageHandler.java:32]: Go ahead and delete, not much need to be backwards-compatible until official 1.0.0

File: jms-core/src/main/java/io/micronaut/jms/listener/JMSListener.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+                    } catch (JMSException e) {
+                        LOGGER.error("Failed to handle successful message receive", e);
+                        ex.addSuppressed(e);
+                    }
+                });
+                if (ex.getSuppressed().length > 0) {
+                    throw ex;
+                }
+            } catch (Throwable e) {
+                errorHandlers.forEach(handler -> handler.handle(session, msg, e));
```
Comment (burtbeckwith) [inline jms-core/src/main/java/io/micronaut/jms/listener/JMSListener.java:120]: should just move this to line 117 in place of `throw ex`, there's a non-zero cost to throwing exceptions (filling in the stack trace, etc.)

File: jms-core/src/main/java/io/micronaut/jms/listener/TransactionalJMSListenerSuccessHandler.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import javax.jms.Message;
+import javax.jms.Session;
+
+/***
+ * Commits a JMS transaction upon successful handling of a message.
+ *
+ * @author Elliott Pope
+ * @since 1.0.0.M2
+ */
+public class TransactionalJMSListenerSuccessHandler implements JMSListenerSuccessHandler {
```
Comment (burtbeckwith) [inline jms-core/src/main/java/io/micronaut/jms/listener/TransactionalJMSListenerSuccessHandler.java:28]: Might want these to be ordered, I assume this one would need to be last or near last, right?

File: jms-core/src/main/java/io/micronaut/jms/templates/JmsProducer.java [hunk 4]
```diff
@@ -91,10 +90,19 @@ public JmsProducer(JMSDestinationType type,
     public void send(@NonNull String destination,
                      @NonNull T body,
                      MessageHeader... headers) {
+
```
Comment (burtbeckwith) [inline jms-core/src/main/java/io/micronaut/jms/templates/JmsProducer.java:93]: Better to join the headers once, e.g.

```
String joinedHeaders = null;
if (LOGGER.isDebugEnabled()) {
    joinedHeaders = Arrays.stream(headers).map(MessageHeader::toString).collect(Collectors.joining(","));
}
```

File: jms-core/src/main/java/io/micronaut/jms/listener/AcknowledgingJMSListenerSuccessHandler.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+/***
+ *
+ * Success handler to acknowledge that the listener has received and processed the message successfully.
+ *  Handler should be enabled if the listener uses the {@link Session#CLIENT_ACKNOWLEDGE} mode.
+ *
+ * @author Elliott Pope
+ * @since 1.0.0.M2
+ */
+public class AcknowledgingJMSListenerSuccessHandler implements JMSListenerSuccessHandler {
```
Comment (burtbeckwith) [inline jms-core/src/main/java/io/micronaut/jms/listener/AcknowledgingJMSListenerSuccessHandler.java:34]: I'm not sure how useful this is - using client ack implies that the user will ack or not along with other use case-specific logic when processing the message. Having it here in the framework is essentially the same as auto acknowledge

**Comment by graemerocher** [comment]
@elliottpope are you interested in continuing this work?

**Comment by graemerocher** [comment]
Merged here https://github.com/micronaut-projects/micronaut-jms/commit/20c6fd72fbab5ddd9330180b891a0938b7b8ba63

**Pull request closed without merging**